src/Halogen/HTML/Core.purs

```diff
@@ -0,0 +1,209 @@
+--| The core types and smart constructors for the HTML DSL.
+module Halogen.HTML.Core
+  ( HTML(..)
+  , element
+  , install
+
+  , Prop(..)
+  , PropF(..)
+  , HandlerF(..)
+  , prop
+  , handler
+
+  , IsProp
+  , toPropString
+
+  , Namespace()
+  , namespace
+  , runNamespace
+
+  , TagName()
+  , tagName
+  , runTagName
+
+  , PropName()
+  , propName
+  , runPropName
+
+  , AttrName()
+  , attrName
+  , runAttrName
+
+  , EventName()
+  , eventName
+  , runEventName
+
+  , ClassName()
+  , className
+  , runClassName
+  ) where
+
+import Prelude
+
+import Data.Bifunctor (Bifunctor, rmap)
+import Data.DOM.Simple.Types (HTMLElement())
+import Data.Exists (Exists(), mkExists)
+import Data.ExistsR (ExistsR(), mkExistsR, runExistsR)
+import Data.Maybe (Maybe(..))
+import Data.Traversable (traverse)
+import Data.Tuple (Tuple(..))
+
+import Halogen.HTML.Events.Handler (EventHandler())
+import Halogen.HTML.Events.Types (Event())
+
+-- | An initial encoding of HTML nodes.
+data HTML p i
+  = Text String
+  | Element (Maybe Namespace) TagName (Array (Prop i)) (Array (HTML p i))
+  | Placeholder p
+
+instance bifunctorHTML :: Bifunctor HTML where
+  bimap f g = go
+    where
+    go (Text s) = Text s
+    go (Element ns name props els) = Element ns name ((g <$>) <$> props) (go <$> els)
+    go (Placeholder p) = Placeholder (f p)
+
+instance functorHTML :: Functor (HTML p) where
+  map = rmap
+
+element :: forall p i. TagName -> Array (Prop i) -> Array (HTML p i) -> HTML p i
+element = Element Nothing
+
+install :: forall p p' i i' m. (Applicative m) => (p -> m (HTML p' i')) -> (i -> i') -> HTML p i -> m (HTML p' i')
+install _ _ (Text s) = pure $ Text s
+install f g (Element ns name props els) = Element ns name ((g <$>) <$> props) <$> traverse (install f g) els
+install f _ (Placeholder p) = f p
+
+-- | A property can be:
+-- | - A JavaScript property for an element (typed, and may not have a
+-- |   corresponding attribute).
+-- | - A raw attribute for an element (stringly typed, will be added directly to
+-- |   the rendered element)
+-- | - A key value used for hinting when diffing HTML.
+-- | - An event handler.
+-- | - A initializer that is triggered once the element for the property has
+-- |   been added to the DOM.
+-- | - A finalizer that is triggered when the element for the property is
+-- |   removed from the DOM.
+data Prop i
+  = Prop (Exists PropF)
+  | Attr (Maybe Namespace) AttrName String
+  | Key String
+  | Handler (ExistsR (HandlerF i))
+  | Initializer (HTMLElement -> i)
```

Comment:
I like the new initializers and finalizers, should eliminate the need for some post render hooks.